### ISSUE TITLE
no-unnecessary-condition: false positive comparing branded intersection types with string literals

### ISSUE THREAD

**Issue by dominik-stagtech** [oxc-project/tsgolint#815]
## Description

`no-unnecessary-condition` reports a false positive when comparing a branded type (intersection of `string` with an object type) against a string literal. It claims the types have no overlap, but at runtime the comparison works correctly.

## Minimal reproduction

```typescript
type Brand<T, B extends string> = T & { readonly $brand: { [K in B]: true } };
type IsoDateString = Brand<string, "IsoDateString">;

declare const date: IsoDateString;

// ❌ false positive: "This condition will always return the same value since the types have no overlap"
if (date === "2025-01-01") {}
```

## Expected behavior

No diagnostic — `IsoDateString` is structurally a `string` at runtime, so comparing it to a string literal is valid and useful.

## Actual behavior

```
typescript-eslint(no-unnecessary-condition): This condition will always return the same value since the types have no overlap.
  Type: "2025-01-01"
  Type: string & $brand<"IsoDateString">
```

## Root cause

`typesHaveOverlap` checks assignability in both directions:
- `isTypeAssignableTo("2025-01-01", string & { $brand: ... })` → false (literal lacks the brand property)
- `isTypeAssignableTo(string & { $brand: ... }, "2025-01-01")` → false (branded type is wider)

Neither succeeds, so it reports no overlap. But the `string` constituent of the intersection clearly overlaps with a string literal.

The fix would be to decompose intersection types before checking overlap — if any constituent type overlaps with the other side, the overall intersection overlaps too.

Note: upstream typescript-eslint doesn't have this issue because it doesn't implement its own overlap check for equality comparisons — it defers to the TypeScript compiler.

## Context

This pattern is common with Zod branded types (`z.string().brand<"IsoDateString">()`) and other nominal typing patterns in TypeScript.

## Versions

- oxlint: 1.55.0
- oxlint-tsgolint: 0.16.0

**Comment by camc314** [comment]
fixed in https://github.com/oxc-project/tsgolint/pull/809

And released in 0.17.2
